ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import tech.pegasys.teku.core.operationvalidators.OperationStateTransitionValidator;
+import tech.pegasys.teku.datastructures.operations.AttesterSlashing;
+import tech.pegasys.teku.datastructures.operations.ProposerSlashing;
+import tech.pegasys.teku.datastructures.operations.SignedVoluntaryExit;
+import tech.pegasys.teku.datastructures.state.BeaconState;
+import tech.pegasys.teku.ssz.SSZTypes.SSZList;
+import tech.pegasys.teku.ssz.SSZTypes.SSZMutableList;
+import tech.pegasys.teku.util.config.Constants;
+
+public class OperationPool<T> {
+
+  private static Map<Class<?>, Integer> maxNumberOfElementsInBlock =
+      Map.of(
+          SignedVoluntaryExit.class, Constants.MAX_VOLUNTARY_EXITS,
+          ProposerSlashing.class, Constants.MAX_PROPOSER_SLASHINGS,
+          AttesterSlashing.class, Constants.MAX_ATTESTER_SLASHINGS);
+
+  private Set<T> operations = new HashSet<>();
+  private OperationStateTransitionValidator<T> operationValidator;
+  private Class<T> clazz;
+
+  public OperationPool(Class<T> clazz, OperationStateTransitionValidator<T> operationValidator) {
+    this.clazz = clazz;
+    this.operationValidator = operationValidator;
+  }
+
+  public SSZList<T> getItemsForBlock(BeaconState stateAtBlockSlot) {
+    SSZMutableList<T> itemsToPutInBlock =
+        SSZList.createMutable(clazz, maxNumberOfElementsInBlock.get(clazz));
+    Iterator<T> iter = operations.iterator();
+    int count = 0;
+    int numberOfElementsToGet = maxNumberOfElementsInBlock.get(clazz);
+    while (count < numberOfElementsToGet && iter.hasNext()) {
+      T item = iter.next();
+      if (operationValidator.validate(stateAtBlockSlot, item).isEmpty()) {
+        itemsToPutInBlock.add(item);
+        count++;
+      }
+      iter.remove();
```

Comment:
Do you want to remove the element even if it wasn't included?  What if its valid for a different fork?

Reply:
I think we would deal with elements that are valid for a different fork in a follow-up PR since there are a good amount of places in the codebase we need to do that.

Reply:
But these items we're dropping won't be recoverable right?  Suggested making the `operations` limited in size, in which case there shouldn't be a problem holding on to these.

Reply:
That's a fair point.

Reply:
Making the change.

Reply:
Or maybe it would be better to toss them into a limited reorg collection or something so we don't keep re-processing them ... In which case we could just leave it for a follow-up

Reply:
Done.

Reply:
I set the operations to 1000 for each pool, which shouldn't be too much of a processing issue. But I don't have any strong preference. Want me to drop the operations or keep them?

Reply:
Maybe it's better to just come back to the reorg handling - lets make sure there's a ticket for reorg handling that links to this line of code.